### PR TITLE
Split Rust compilation into its own CLI bin

### PR DESCRIPTION
**Pull request by leonardoalt**
As suggested by @lvella .
I'm not too sure about this actually. I agree that compiling Rust -> powdr-asm should have a separate pipeline somewhere, but before we could do Rust -> proof in a single command, and with this we can't anymore.

- Do we expose the `prove` command and everything else here as well? That'd be a lot of duplicated code.
- Do we force the user to first run `powdr-rs compile <rust_proj>` then `powdr pil/prove ...`? That's potentially worse UX than right now.
- Do we keep all the Rust stuff in the `cli` crate?

### CODE REVIEW

**Comment by lvella** [comment]
If we do it like C/C++/Rust compilers, `powdr-riscv` can simply call the `powdr` binary for the user.

But yes, I can see some code duplication from parsing/forwarding the `powdr` options taken from `powdr-riscv`, but we can also just pass the options we don't recognize dumbly to `powdr`.

File: Cargo.toml [hunk 1]
```diff
@@ -8,6 +8,7 @@ members = [
     "number",
     "parser",
     "cli",
+	"cli-rs",
```
Comment (lvella) [inline Cargo.toml:11]: What is a tab? A miserably little pile of spaces?

**Comment by leonardoalt** [comment]
>  but we can also just pass the options we don't recognize dumbly to powdr.

Yea that's true, though quite ugly to do that from Rust. I don't see a better option

**Comment by leonardoalt** [comment]
This is ready for review.

Currently it requires 2 commands to go from Rust -> proof, I think it's not too bad actually. I wanted to keep it minimal and try it out first like this before doing the whole args forwarding etc.

File: cli-rs/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+                    .unwrap();
+            if continuations {
+                runtime.with_poseidon()
+            } else {
+                runtime
+            }
+        }
+        (None, false) => powdr_riscv::Runtime::base(),
+        (None, true) => powdr_riscv::Runtime::base().with_poseidon(),
+    };
```
Comment (georgwiese) [inline cli-rs/src/main.rs:261]:
```suggestion    let mut runtime = match coprocessors {
        Some(list) => {
            powdr_riscv::Runtime::try_from(list.split(',').collect::<Vec<_>>().as_ref()).unwrap()
        }
        None => powdr_riscv::Runtime::base(),
    };
    if continuations && !runtime.has_submachine("poseidon_gl") {
        runtime = runtime.with_poseidon();
    }
```

Cleaner IMO, and also fixed passing `-c --coprocessors poseidon_gl`, which previously complained that the poseidon coprocessor is added twice.

File: cli-rs/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+
+    let (_asm_file_path, _asm_contents) =
+        powdr_riscv::compile_rust::<F>(file_name, output_dir, true, &runtime, continuations)
+            .ok_or_else(|| vec!["could not compile rust".to_string()])?;
```
Comment (georgwiese) [inline cli-rs/src/main.rs:265]:
```suggestion
    powdr_riscv::compile_rust::<F>(file_name, output_dir, true, &runtime, continuations)
        .ok_or_else(|| vec!["could not compile rust".to_string()])?;
```

**Review by georgwiese — COMMENTED** [review]
Cool!

For reference, this is what needs to be done to get the `many_chunks` example working on the command line:

I tried:
```bash
# Writes output/many_chunks.asm
powdr-rs compile riscv/tests/riscv_data/many_chunks -o output -c
# Writes output/chunk_*/commits.bin files
powdr-rs execute output/many_chunks.asm -o output -c --witness
# Computes proof for chunk 0
powdr prove output/many_chunks.asm -d output/chunk_0 --backend estark-sta
rky
```

Should we document this in the book? Continuations are experimental and not sound yet, but for someone just running some quick benchmarks it could still be useful?

I also think it's a bit weird that `powdr-rs execute` does witness generation. I *think* we could also have `powdr-rs execute` persist a CSV file containing the bootloader inputs that can be passed to `powdr pil --witness-values`. That could be a cleaner separation?

File: book/src/installation.md [hunk 4]
```diff
@@ -38,3 +45,21 @@ cd powdr
 # install powdr-cli
 cargo install --features halo2 --path ./cli
 ```
+
+## Building *powdr-rs*
+
+Using a single Cargo command (enable the Halo2 backend to use it with the cli):
```
[the remark this answers is not in the document]
Comment (leonardoalt) [reply book/src/installation.md:51]: true!

**Comment by leonardoalt** [comment]
I'll add something to the book!

> I also think it's a bit weird that powdr-rs execute does witness generation. I think we could also have powdr-rs execute persist a CSV file containing the bootloader inputs that can be passed to powdr pil --witness-values. That could be a cleaner separation?

We could do that too, I see pros and cons:
- powdr-rs running witgen definitely crosses a border
- however it's more automated than \/
- we need to make the user aware of yet another artifact (the bootloader csv inputs) which they otherwise wouldn't need to care
- how would the user continue after that? Would it work for the user to just call `powdr` with `--witness-values` passing the bootloader inputs?

**Comment by leonardoalt** [comment]
We decided to keep it as is for now and refactor the old cli first

File: book/src/frontends/zk-continuations.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+For the details of how memory is handled in the ZK-continuations case please
+see [this](https://github.com/powdr-labs/powdr/issues/814).
+
+`powdr-rs` has experimental support to ZK-continuations, which can be used as follows.
+
+Let's use as example test `many chunks` from the `riscv` crate:
+
+```rust
+{{#include ../../../riscv/tests/riscv_data/many_chunks/src/lib.rs}}
```
Comment (georgwiese) [inline book/src/frontends/zk-continuations.md:16]: Hehe, actually the comments say that it tries not to access any memory, we should probably change it to a more realistic example in an upcoming PR!

**Review by georgwiese — APPROVED** [review]
Nice!

**Pull request merged**